Source/Core/VideoCommon/TmemBase.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <array>
+
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/TmemBase.h"
+
+namespace TmemBase
+{
+// On TMEM configuration changed:
+// 1. invalidate stage.
+
+struct texture_unit_state
+{
+  enum State
+  {
+    INVALID, // Configuration has changed
+    VALID, // But not cached due to either being too big, or overlapping with another valid texture unit
+    CACHED,
+  };
+
+  TexImage1 even;
+  TexImage2 odd;
+  State state;
+
+  int even_base;
+  int even_size;
+  int odd_base;
+  int odd_size;
+};
+static std::array<texture_unit_state, 8> s_unit;
+
+void ConfigurationChanged(int bp_addr, int config)
+{
+  // Extract bits encoding texture unit
+  u8 unit = ((bp_addr & 0x20) >> 3) | (bp_addr & 3);
+  bool even = bp_addr < BPMEM_TX_SETIMAGE2;
+
+  texture_unit_state& unit_state = s_unit[unit];
+
+  if (even)
+  {
+    unit_state.even.hex = config;
+  }
+  else
+  {
+    unit_state.odd.hex = config;
+  }
+  unit_state.state = texture_unit_state::INVALID;
+}
+
+// On invalidate cache
+// 1. invalidate all
+
+void Invalidate(u32 param)
+{
+  // The excat arguments of Invalidate commands is currently unknown.
+  // It appears to contain the TMEM address and a size.
+
+  // For simplicity, we will just invalidate everything
+  for (auto& unit : s_unit)
+  {
+    unit.state = texture_unit_state::INVALID;
+  }
+}
+
+// On bind:
+// 1. store the memory region(s) bound
+// 2. if texture size is small enough to fit in region AND there is no overlap, mark as cached.
+// 3. If there is cached overlap, invalidate.
+
+static inline u32 calculate_unit_size(TexImage1 config)
+{
+  // These are the only cache sizes supported by the sdk
+  if (config.cache_width == config.cache_height)
+  {
+    switch (config.cache_width)
+    {
+    case 3:  // 32KB
+      return 32 * 1024;
+    case 4:  // 128KB
+      return 128 * 1024;
+    case 5:  // 512KB
+      return 512 * 1024;
+    default:
+      break;
+    }
+  }
+
+  // However, the registers allow a much larger amount of configurablity.
+  // Maybe other sizes are broken?
+  // Until hardware tests are done, this is a guess at the size algorithm
+
+  return 512 * (1 << config.cache_width) * (1 << config.cache_height);
+}
+
+static inline u32 calculate_unit_size(TexImage2 config)
+{
+  TexImage1 casted_config;
+  casted_config.hex = config.hex;
+  return calculate_unit_size(casted_config);
+}
+
+static bool overlap(int a, int a_size, int b, int b_size)
+{
+  return (a <= b && (a + a_size) >= b) || (b <= a && (b + b_size) >= a);
+}
+
+static bool process_overlaps_single(u8 unit_id, int base, int size)
```

Comment:
Would there be any benefit in having a `struct` parameter instead of separate `base`/`size` and reuse the same as `TextureUnitState.even`/`TextureUnitState.odd`? That way you could move the `overlap` method into that struct; potentially as operator if it makes sense; plus get the added benefit of less ambiguity when faced with 2 or 4 `int` parameters (like, is it `base` or `size` first; or are they next to each other in the `overlap` case?)
It could also be the home for `CalculateUnitSize` (and perhaps others) if it makes sense in there.

Reply:
I'm really unhappy with this code and the way implementation details (finding overlaps) is mixed with the higher level concepts.

What I want to do is implement a generic [ interval tree](https://en.m.wikipedia.org/wiki/Interval_tree) implementation in common, which appears to be the correct data structure for storing and checking for overlaps.

Then I can just implement tmem in terms of the interval tree.